plugins/woocommerce/src/Internal/Updates/BatchProcessingController.php

```diff
@@ -0,0 +1,200 @@
+	const CONTROLLER_CRON_NAME = 'wc_schedule_pending_updates';
+	const SINGLE_PROCESS_MIGRATION_ACTION = 'wc_run_action_update';
+	const ACTION_GROUP = 'wc_updates';
+
+	/**
+	 * BatchProcessingController constructor.
+	 *
+	 * Schedules necessary actions to process batches.
+	 */
+	public function __construct() {
+		add_action( self::CONTROLLER_CRON_NAME, array( $this, 'schedule_processes' ) );
+		add_action( self::SINGLE_PROCESS_MIGRATION_ACTION, array( $this, 'process_single_update' ), 10, 2 );
+	}
+
+	/**
+	 * Starts an update.
+	 *
+	 * @param string $updater_class_name Fully qualified class name of an updater, must be child class of `BatchProcessor`.
+	 */
+	public function enqueue_processor( string $updater_class_name ) {
+		$pending_updates = $this->get_pending();
+		if ( ! in_array( $updater_class_name, array_keys( $pending_updates ) ) ) {
+			$pending_updates[] = $updater_class_name;
+			$this->set_pending_processes( $pending_updates );
+		}
+		$this->schedule_init_cron();
+	}
+
+	/**
+	 * Helper method to start update cron.
+	 */
+	private function schedule_init_cron() {
+		as_schedule_single_action(
+			$this->get_start_timestamp_for_init_action(),
+			self::CONTROLLER_CRON_NAME,
+			$this->get_args_for_init_cron(),
+			self::ACTION_GROUP
+		);
+	}
+
+	/**
+	 * Schedules update for all updaters that may be stuck. This method is called in CONTROLLER_CRON_NAME action.
+	 */
+	public function schedule_processes() {
+		$pending_updates = $this->get_pending();
+		if ( empty( $pending_updates ) ) {
+			return;
+		}
+		foreach ( $pending_updates as $update_name ) {
+			if ( ! $this->already_scheduled( $update_name ) ) {
+				$this->schedule_next_batch( $update_name );
+			}
+		}
+		$this->schedule_init_cron();
+	}
+
+	/**
+	 * Process update for a scheduled updater.
+	 *
+	 * @param string $update_name Fully qualified class name of the updater. Must be child class `BatchProcessor`.
+	 */
+	public function process_single_update( string $update_name ) {
+		$updater = $this->get_processor_instance( $update_name );
+		$still_pending = $updater->process_batch();
+		if ( $still_pending ) {
+			$this->schedule_next_batch( $update_name );
+		} else {
+			$this->mark_pending_process_complete( $update_name );
+			$updater->mark_process_complete();
+		}
+	}
+
+	/**
+	 * Scheduler next update batch.
+	 *
+	 * @param string $update_name Fully qualified class name of the updator.
+	 *
+	 * @return int Action ID.
+	 */
+	private function schedule_next_batch( string $update_name ) : int {
+		return as_schedule_single_action( time(), self::SINGLE_PROCESS_MIGRATION_ACTION, array( $update_name ) );
+	}
+
+	/**
+	 * Whether the given updater is already scheduled.
+	 * Differs from `as_has_scheduled_action` in that this excludes actions in progress.
+	 *
+	 * @param string $update_name Fully qualified class name of the updator.
+	 *
+	 * @return bool Whether it's already scheduled.
+	 */
+	public function already_scheduled( string $update_name ) {
```

Comment:
Name change suggestion:

```suggestion
	public function is_enqueued( string $processor_class_name ) {
```

Reply:
Note that there is already a method `is_enqueued` that checks if the process is registered and will be scheduled. This method is an abstraction for `as_has_scheduled_action` which indicates if a batch is going to be processed very soon.
Updated the param name in https://github.com/woocommerce/woocommerce/pull/33233/commits/e922be012e20ff76dc1fc8357d357b573f8c4cd2#diff-8989ad7f50c6e63bf0ff83b5df17d72121166df76e16875ec985062b8ca740d5R185

Reply:
So enqueued !== scheduled, ok, I get it. In that case I'd go with `is_scheduled` for consistency and readability.

Reply:
👍 , renamed in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-77d9531091cce5f175d5bb8fe3a1560d9dda936d370c81fabb90ca0f46633e52R197